File: backend/src/operators.js
// Identification de l'opérateur attributaire d'un numéro, à partir de
// l'open data ARCEP MAJNUM (associe chaque tranche de numéros à
// l'opérateur qui l'a reçue). Permet d'enrichir le lookup : « ce numéro
// appartient à la plage de l'opérateur X », et de repérer les grossistes
// VoIP massivement utilisés par les centres d'appels de démarchage.
//
// Format MAJNUM : EZABPQM;Tranche_Debut;Tranche_Fin;Mnémo;Territoire;Date
// (CSV latin-1, séparateur ';'). Tranches en national 10 chiffres (0XXXXXXXXX).

const MAJNUM_URL = 'https://extranet.arcep.fr/uploads/MAJNUM.csv';

// Opérateurs grossistes / VoIP fréquemment associés au démarchage
// téléphonique (mnémoniques ARCEP → libellé lisible). Non exhaustif ;
// informatif, n'entraîne pas de blocage à lui seul.
const KNOWN_OPERATORS = {
  OXIL: 'Oxilog',
  UBIC: 'Ubicentrex',
  KAVE: 'Kav El International',
  GETG: 'Getg',
  QWLK: 'Quewilk',
  YAAT: 'Yaat',
  FLCA: 'Foliateam',
  MANI: 'Manifone',
  COMU: 'Comunik CRM',
  IPDI: 'IP Directions',
  DVSV: 'Diabolocom',
  OING: 'Ovanet',
  FREE: 'Free',
  SFR0: 'SFR',
  F000: 'Orange',
  BYTL: 'Bouygues Telecom',
};

// Index trié en mémoire : [ [debut, fin, mnemo], ... ] trié par debut.
let ranges = [];
let starts = [];

function buildIndex(text) {
  const out = [];
  const lines = text.split('\n');
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i];
    if (!line) continue;
    const cols = line.split(';');
    if (cols.length < 4) continue;
    const deb = Number.parseInt(cols[1], 10);
    const fin = Number.parseInt(cols[2], 10);
    const mnemo = (cols[3] || '').trim();
    if (!Number.isInteger(deb) || !Number.isInteger(fin) || !mnemo) continue;
    out.push([deb, fin, mnemo]);
  }
  out.sort((a, b) => a[0] - b[0]);
  ranges = out;
  starts = out.map((r) => r[0]);
  return out.length;
}

// Recherche dichotomique de la tranche contenant le numéro national.
function mnemoForNational(nat) {
  let lo = 0;
  let hi = starts.length - 1;
  let ans = -1;
  while (lo <= hi) {
    const mid = (lo + hi) >> 1;
    if (starts[mid] <= nat) {
      ans = mid;
      lo = mid + 1;
    } else {
      hi = mid - 1;
    }
  }
  if (ans >= 0 && ranges[ans][0] <= nat && nat <= ranges[ans][1]) {
    return ranges[ans][2];
  }
  return null;
}

// e164 (+33XXXXXXXXX) → { mnemo, name } ou null. Métropole uniquement.
export function operatorFor(e164) {
  if (typeof e164 !== 'string' || !e164.startsWith('+33')) return null;
  const nat = Number.parseInt('0' + e164.slice(3), 10);
  if (!Number.isInteger(nat)) return null;
  const mnemo = mnemoForNational(nat);
  if (!mnemo) return null;
  return { mnemo, name: KNOWN_OPERATORS[mnemo] || null };
}

export async function refreshOperators() {
  const res = await fetch(MAJNUM_URL, { signal: AbortSignal.timeout(60_000) });
  if (!res.ok) throw new Error(`HTTP ${res.status}`);
  const buf = await res.arrayBuffer();
  if (buf.byteLength > 20_000_000) throw new Error('MAJNUM trop volumineux');
  // ARCEP publie en latin-1.
  const text = new TextDecoder('latin1').decode(buf);
  const n = buildIndex(text);
  if (n === 0) throw new Error('MAJNUM vide');
  return n;
}

export function operatorsLoaded() {
  return ranges.length;
}

File: backend/src/operators.rs
//! Identification de l'opérateur attributaire d'un numéro via l'open data
//! ARCEP MAJNUM (tranche de numéros -> opérateur). Index trié en mémoire,
//! recherche dichotomique.

use std::collections::HashMap;

pub const MAJNUM_URL: &str = "https://extranet.arcep.fr/uploads/MAJNUM.csv";

/// Mnémoniques ARCEP -> libellé lisible (grossistes VoIP souvent liés au
/// démarchage inclus). Informatif, ne déclenche pas de blocage.
fn known_name(mnemo: &str) -> Option<&'static str> {
    Some(match mnemo {
        "OXIL" => "Oxilog",
        "UBIC" => "Ubicentrex",
        "KAVE" => "Kav El International",
        "GETG" => "Getg",
        "QWLK" => "Quewilk",
        "YAAT" => "Yaat",
        "FLCA" => "Foliateam",
        "MANI" => "Manifone",
        "COMU" => "Comunik CRM",
        "IPDI" => "IP Directions",
        "DVSV" => "Diabolocom",
        "FREE" => "Free",
        "SFR0" => "SFR",
        "F000" => "Orange",
        "BYTL" => "Bouygues Telecom",
        _ => return None,
    })
}

#[derive(Clone)]
pub struct Operator {
    pub mnemo: String,
    pub name: Option<String>,
}

#[derive(Default)]
pub struct OperatorIndex {
    // (début, fin) en national i64 (0XXXXXXXXX) -> mnémo, trié par début.
    ranges: Vec<(i64, i64, String)>,
}

impl OperatorIndex {
    pub fn len(&self) -> usize {
        self.ranges.len()
    }

    /// Reconstruit l'index depuis le CSV MAJNUM (latin-1 déjà décodé).
    pub fn build(text: &str) -> Self {
        let mut ranges = Vec::new();
        for line in text.lines().skip(1) {
            let cols: Vec<&str> = line.split(';').collect();
            if cols.len() < 4 {
                continue;
            }
            let (Ok(deb), Ok(fin)) = (cols[1].trim().parse::<i64>(), cols[2].trim().parse::<i64>())
            else {
                continue;
            };
            let mnemo = cols[3].trim();
            if mnemo.is_empty() {
                continue;
            }
            ranges.push((deb, fin, mnemo.to_string()));
        }
        ranges.sort_by_key(|r| r.0);
        OperatorIndex { ranges }
    }

    fn mnemo_for_national(&self, nat: i64) -> Option<&str> {
        // dernière tranche dont le début <= nat
        let idx = self.ranges.partition_point(|r| r.0 <= nat);
        if idx == 0 {
            return None;
        }
        let (deb, fin, mnemo) = &self.ranges[idx - 1];
        if *deb <= nat && nat <= *fin {
            Some(mnemo)
        } else {
            None
        }
    }

    /// e164 (+33XXXXXXXXX) -> opérateur (métropole).
    pub fn operator_for(&self, e164: &str) -> Option<Operator> {
        let rest = e164.strip_prefix("+33")?;
        let nat: i64 = format!("0{rest}").parse().ok()?;
        let mnemo = self.mnemo_for_national(nat)?;
        Some(Operator {
            mnemo: mnemo.to_string(),
            name: known_name(mnemo).map(str::to_string),
        })
    }

    /// Réputation : combien de numéros distincts d'une liste appartiennent à
    /// chaque opérateur, trié décroissant.
    pub fn reputation(&self, numbers: &[String]) -> Vec<(String, Option<String>, i64)> {
        let mut counts: HashMap<String, i64> = HashMap::new();
        for num in numbers {
            if let Some(op) = self.operator_for(num) {
                *counts.entry(op.mnemo).or_insert(0) += 1;
            }
        }
        let mut out: Vec<_> = counts
            .into_iter()
            .map(|(m, c)| {
                let name = known_name(&m).map(str::to_string);
                (m, name, c)
            })
            .collect();
        out.sort_by_key(|x| std::cmp::Reverse(x.2));
        out
    }
}

/// Télécharge et parse MAJNUM. `text` récupéré en latin-1 par l'appelant.
pub async fn fetch_majnum() -> Result<String, String> {
    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(60))
        .build()
        .map_err(|e| e.to_string())?;
    let resp = client
        .get(MAJNUM_URL)
        .send()
        .await
        .map_err(|e| e.to_string())?;
    if !resp.status().is_success() {
        return Err(format!("HTTP {}", resp.status()));
    }
    let bytes = resp.bytes().await.map_err(|e| e.to_string())?;
    if bytes.len() > 20_000_000 {
        return Err("MAJNUM trop volumineux".into());
    }
    // ARCEP publie en latin-1 (ISO-8859-1) : conversion octet -> char.
    Ok(bytes.iter().map(|&b| b as char).collect())
}
